### PR TITLE
[GlobalIsel] Modernize truncate of ext.

### PR DESCRIPTION
**Pull request by tschuett**
Credits:
https://github.com/llvm/llvm-project/pull/90964
https://reviews.llvm.org/D87050

combine-trunc.mir

Functional changes intended.

### CODE REVIEW

File: llvm/lib/CodeGen/GlobalISel/CombinerHelperCasts.cpp [hunk 1]
```diff
@@ -113,3 +113,45 @@ bool CombinerHelper::matchNonNegZext(const MachineOperand &MO,
...
+  Register Src = Ext->getSrcReg();
+  LLT DstTy = MRI.getType(Dst);
+  LLT SrcTy = MRI.getType(Src);
+
+  if (SrcTy == DstTy) {
+    // The source and the destination are equally sized. We need to copy.
+    MatchInfo = [=](MachineIRBuilder &B) { B.buildCopy(Dst, Src); };
+
+    return true;
+  } else if (SrcTy.getScalarSizeInBits() < DstTy.getScalarSizeInBits()) {
```
[the remark this answers is not in the document]
Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelperCasts.cpp:136]: I always hoped `else if` counts as an excuse, but:
https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return
Nope.

File: llvm/lib/CodeGen/GlobalISel/CombinerHelperCasts.cpp [hunk 2]
```diff
@@ -113,3 +113,45 @@ bool CombinerHelper::matchNonNegZext(const MachineOperand &MO,
...
+    // Protect against scalable vectors.
```
[the remark this answers is not in the document]
Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelperCasts.cpp:137]: The original code used `SrcTy.getSizeInBits()`, which asserts for scalable vectors. The comment is about `getScalarSizeInBits` .

[the remark this answers is not in the document]
Comment (arsenm) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelperCasts.cpp:137]: That makes it make less sense? getScalarSizeInBits should still work for scalable

[the remark this answers is not in the document]
Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelperCasts.cpp:137]: I removed the comment. Wait a minute. `getScalarSizeInBits` works with all kinds of shapes.

**Review by arsenm — APPROVED** [review]

File: llvm/lib/CodeGen/GlobalISel/CombinerHelperCasts.cpp [hunk 3]
```diff
@@ -149,7 +149,7 @@ bool CombinerHelper::matchTruncateOfExt(const MachineInstr &Root,
   }
 
   if (SrcTy.getScalarSizeInBits() > DstTy.getScalarSizeInBits()) {
-    // The source is larger than the destination. We need to truncate.
+    // If the  source is larger than the destination, we need to truncate.
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelperCasts.cpp:152]: Extra space added between words

**Pull request merged**